### PR TITLE
Hex bin digit grouping

### LINKED ISSUE #2538: Underscores only every 2^n digits in non-decimal numbers
What it does: Warns if an integral or floating point **non-decimal** number is split somewhere other than between n^2 and (n^2)+1 (for constant n≥1). Hexadecimal, octal and binary numbers are used when dealing with 2^n fractions or multiples of bytes.

Examples:

- `0xabc_def`. Probably what was meant was `0xab_cdef`, `0xab_cd_ef` or there's a missing nibble in the second half.
- `0xa_b_c_d_e_f` is more verbose but no clearer than `0xabcdef`.

### PR DESCRIPTION
**Pull request by cgm616**
This revives and updates an old pr (#3391) for the current API. 

Closes #2538.

---

*Please keep the line below*
changelog: Add [`unusual_byte_groupings`] lint.

### CODE REVIEW

**Comment by cgm616** [comment]
I think this PR is all set now. There was one change made with the suggested grouping for hexadecimal literals that might need discussion. Instead of groups of 4, like `0xDEAD_BEEF`, clippy will now suggest groups of two, like `0xDE_AD_BE_EF`. This is to make sure that clippy's fixes won't contravene this lint, which suggests byte-sized chunks for hex literals.

File: clippy_lints/src/literal_representation.rs [hunk 1]
```diff
@@ -82,6 +82,25 @@ declare_clippy_lint! {
     "integer literals with digits grouped inconsistently"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Warns if a hexadecimal or binary literal is not grouped
```
Comment (flip1995) [inline clippy_lints/src/literal_representation.rs:86]:
```suggestion
    /// **What it does:** Warns if a hexadecimal or binary literals are not grouped
```

**Review by flip1995 — COMMENTED** [review]
I really don't like the suggestion change of the other lint. IMO splitting hex numbers in groups of 4 is more accepted than groups of 2.

**Comment by cgm616** [comment]
> I really don't like the suggestion change of the other lint. IMO splitting hex numbers in groups of 4 is more accepted than groups of 2.

Yeah, that's fair. I just reverted that change and switched this new lint to checking if hex literals are in groups of four to stay consistent.

File: tests/ui/literals.rs [hunk 2]
```diff
@@ -7,10 +7,10 @@
...
 fn main() {
     let ok1 = 0xABCD;
-    let ok3 = 0xab_cd;
-    let ok4 = 0xab_cd_i32;
-    let ok5 = 0xAB_CD_u32;
-    let ok5 = 0xAB_CD_isize;
+    let ok3 = 0xabcd;
+    let ok4 = 0xabcd_i32;
+    let ok5 = 0xABCD_u32;
+    let ok5 = 0xABCD_isize;
```
Comment (flip1995) [inline tests/ui/literals.rs:13]: We should not lint for groupings of 2 in hex literals either.

File: tests/ui/literals.stderr [hunk 3]
```diff
@@ -69,5 +69,25 @@ error: digits grouped inconsistently by underscores
...
+error: digits of hex or binary literal not grouped by four
+  --> $DIR/literals.rs:38:18
+   |
+LL |     let fail25 = 0b01_100_101;
+   |                  ^^^^^^^^^^^^ help: consider: `0b0110_0101`
+
+error: digits of hex or binary literal not grouped by four
+  --> $DIR/literals.rs:39:16
+   |
+LL |     let ok26 = 0x6_A0_BF;
```
Comment (flip1995) [inline tests/ui/literals.stderr:89]: This is definitely a FP

**Review by flip1995 — CHANGES_REQUESTED** [review]

File: tests/ui/literals.stderr (hunk 3 above)
Comment (cgm616) [reply tests/ui/literals.stderr:89]: Sorry, FP?

Comment (ebroto) [reply tests/ui/literals.stderr]: FP is an abbreviation of false positive

Comment (cgm616) [reply tests/ui/literals.stderr]: I think I have fixed that, now.

**Comment by cgm616** [comment]
For some reason, my local builds always fail with seven or so of this error: 

```
no variant or associated item named `ConstBlock` found for enum `rustc_ast::ExprKind`
```

Does anyone know why? It makes debugging this lint really hard since I can't run any of the tests.

**Comment by ebroto** [comment]
>my local builds always fail with seven or so of this error

I think you need to run `setup-toolchain.sh` again to catch up with the supported rustc version. See `doc/basics.md`.

**Comment by cgm616** [comment]
> > my local builds always fail with seven or so of this error
> 
> I think you need to run `setup-toolchain.sh` again to catch up with the supported rustc version. See `doc/basics.md`.

That helps partially; thank you. Now, though, I have a bunch of errors about crates being built with incompatible versions of `rustc`. I've tried deleting the `target/` folder and running `cargo clean`, but neither work.

**Comment by ebroto** [comment]
>I have a bunch of errors about crates being built with incompatible versions of rustc

Could you share the errors?

Also, make sure you rebased on top of the current master.

**Comment by cgm616** [comment]
> Could you share the errors?

They went away mysteriously after waiting a while. It's possible that rust-analyzer running in VSCode had something to do with it. Thanks for the help!

**Comment by flip1995** [comment]
> It's possible that rust-analyzer running in VSCode had something to do with it.

I have the `    "rust-analyzer.checkOnSave.extraArgs": ["--target-dir", "target/rust-analyzer"],` config option set for rust-analyzer (I use coc.nvim, but I'm sure there's something similar for VSCode). This will create a `rust-analyzer` subdir in the target dir (for every project), so no more conflicting `cargo build` and `rust-analyzer` builds. I only suggest to use that if you have enough memory to spare.

File: tests/ui/literals.rs (hunk 2 above)
Comment (flip1995) [inline tests/ui/literals.rs:13]: These changes to the test file should be unnecessary now?

File: clippy_lints/src/literal_representation.rs [hunk 4]
```diff
@@ -82,6 +82,25 @@ declare_clippy_lint! {
...
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let x: u32 = 0xFFF_FFF;
+    /// let y: u8 = 0b01_011_101;
+    /// ```
+    pub UNUSUAL_BYTE_GROUPING,
```
Comment (flip1995) [inline clippy_lints/src/literal_representation.rs:99]:
```suggestion
    pub UNUSUAL_BYTE_GROUPINGS,
```

Make sure to run `cargo dev update_lints` and to update the reference files after this change.

**Review by flip1995 — COMMENTED** [review]
Two small things left, before this is ready to get merged.

File: tests/ui/literals.rs (hunk 2 above)
Comment (cgm616) [reply tests/ui/literals.rs]: Yep!

**Review by flip1995 — APPROVED** [review]

**Comment by flip1995** [comment]
Thanks! Waiting on rustup.

@cgm616 I removed the `WIP` from the title. Was there something you thought was still missing?

**Comment by cgm616** [comment]
> @cgm616 I removed the `WIP` from the title. Was there something you thought was still missing?

Nope! Just leftover from the beginning. Thanks for everything!

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**